src/Pipeline/Queue/Atomic.php

```diff
@@ -0,0 +1,215 @@
+<?php
+
+/*
+ * This file is part of the Predis package.
+ *
+ * (c) Daniele Alessandri <suppakilla@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Predis\Pipeline\Queue;
+
+use SplQueue;
+use Throwable;
+use Traversable;
+use NoRewindIterator;
+use InvalidArgumentException;
+use Predis\Command\CommandInterface;
+use Predis\Command\RawCommand;
+use Predis\Connection\ConnectionInterface;
+use Predis\Response\ErrorInterface;
+
+/**
+ * Atomic command queue that wraps flushes into MULTI / EXEC transactions.
+ */
+class Atomic implements CommandQueueInterface
+{
+    private $discarded = false;
+    /** @var CommandQueueInterface */
+    private $innerQueue;
+    /** @var SplQueue */
+    private $txCommands;
+
+    /**
+     * @inheritdoc
+     */
+    public function __construct(CommandQueueInterface $queue)
+    {
+        $this->checkInnerQueue($queue);
+
+        $this->innerQueue = $queue;
+        $this->txCommands = new SplQueue;
+    }
+
+    /**
+     * Verifies if the inner command queue is compatibile with atomic queues.
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function checkInnerQueue(CommandQueueInterface $queue): void
+    {
+        if ($queue instanceof static) {
+            throw new InvalidArgumentException('Atomic command queues cannot be nested');
+        }
```

Comment:
This suggestion can be skipped but we can return early on the happy path as it makes the body of the method look more useful.

```suggestion
        if (! $queue instanceof static) {
            return;
        }
        throw new InvalidArgumentException('Atomic command queues cannot be nested');
```

Reply:
I understand the reasoning behind your suggestion. Actually I implemented the body of the method that way because I think we may introduce a few more checks later (I have a vague idea for more specialized command queues), in that case we would end up having a few cascading checks. Will leave it as committed for now, we'll see later.

Reply:
Ok, that makes sense.